### ISSUE TITLE
Polymer flooding does not work as expected when keywords like PORV/NNC are used

### ISSUE THREAD

**Issue by Wenyue-Sun** [OPM/opm-simulators#3386]
I'm running some polymer simulations with a model that is specified with keywords PORV, TRNAX/Y/Z, and non-neighborhood connection (NNC). Found that the simulation does not converge with polymer flooding turned on when PORV is used, or NNC is used. Attached zip file contains the following models:
1. 2D_POLYMER_INJECTIVITY.DATA
  - test model from opm-test repo, simulation success
2. 2D_POLYMER_INJECTIVITY_NNC.DATA
  - TRANZ is set to zero, and NNC is set for last layer to the 3rd layer. Everything else is the same. simulator unconverged
3. 2D_POLYMER_INJECTIVITY_PORV.DATA
  - PORV and TRANX/Y/Z are used
  - DX/DY/DZ is changed from original value (e.g.; 25) to 1. Simulation unconverged.
  - (I tested the case that DX/DY/DZ is unchanged from original values, the simulation converged, so I guess when polymer option is turned on, the code is still using DX/DY/DZ instead of the values provided by PORV/TRANX)

Thanks for your attention, any help/suggestions here would be appreciated.

[polymer_test.zip](https://github.com/OPM/opm-simulators/files/6665770/polymer_test.zip)

**Comment by GitPaean** [comment]
Hi, Wenyue, 

Welcome to OPM community. 

I did a quick test with the master branch, OPM-flow ran through the example 2D_POLYMER_INJECTIVITY_NNC.DATA and suffered from convergence issue for example 2D_POLYMER_INJECTIVITY_PORV.DATA. 

Indeed, in the polymer code there is some geometric information of the grid cell used in the simulation, especially when we do the polymer injectivity simulation.

Before I look into further, can you tell me more about what kinds of simulation you are looking for. The polymer injectivity simulator is a simulator designed by one of our clients intended to simulate the buildup of the skin affects during polymer and water flooding. We designed special new keywords for this functionality. 

If you only need to do some general polymer flooding simulation, you can also look at the examples at the folders polymer_simple2D and polymer_oilwater in the opm-tests repo (https://github.com/OPM/opm-tests). I thought we had more examples in 3D or so in the repo, but I did not see them just now. If you are interested, I can find some for you. 

Just let me know.

**Comment by Wenyue-Sun** [comment]
> Hi, Wenyue,
> 
> Welcome to OPM community.
> 
> I did a quick test with the master branch, OPM-flow ran through the example 2D_POLYMER_INJECTIVITY_NNC.DATA and suffered from convergence issue for example 2D_POLYMER_INJECTIVITY_PORV.DATA.
> 
> Indeed, in the polymer code there is some geometric information of the grid cell used in the simulation, especially when we do the polymer injectivity simulation.
> 
> Before I look into further, can you tell me more about what kinds of simulation you are looking for. The polymer injectivity simulator is a simulator designed by one of our clients intended to simulate the buildup of the skin affects during polymer and water flooding. We designed special new keywords for this functionality.
> 
> If you only need to do some general polymer flooding simulation, you can also look at the examples at the folders polymer_simple2D and polymer_oilwater in the opm-tests repo (https://github.com/OPM/opm-tests). I thought we had more examples in 3D or so in the repo, but I did not see them just now. If you are interested, I can find some for you.
> 
> Just let me know.

Hi, Kai, 

Thanks for the quick reply and the explanation. I've tried the polymer_oilwater example as well and seems there is still similar unexpected behavior. I attached the simulation files below (case w/wo NNc, w/wo PORV), with some results I got using a commercial simulator. 

I'm doing some research for general polymer flooding experiments using some surrogate techniques, similar to flownet, that's why I'm playing with these PORV/TRANS/NNC keywords. Any help is appreciated. Thanks!


[poly_oilwater.zip](https://github.com/OPM/opm-simulators/files/6671812/poly_oilwater.zip)

**Comment by GitPaean** [comment]
Hi, Wenyue, 

Thanks for sharing the case and the result. And I confirmed your finding. I do not know the cause yet. I will have a look as soon as possible.

**Comment by GitPaean** [comment]
Hi, Wenyue. 

Sorry for not getting back to you sooner. I took a brief look of your example 2D_OILWATER_POLYMER_USE_PORV.DATA . 

In this example, the cell volume is 1, and you assigned pore volume 50000, which will results in porosity of 50000.  It depends how the simulator is formulated, it might or might not cause big problem.  It depends in details how your equation is formulated and how the geometry information is used. It causes problem for OPM-flow, but I could not find enough time to dig in for this. 

For this example, I suggest you increase the grid cell volume, then make sure the pore volume is smaller than the grid cell volume, so you have a porosity smaller than 1, which is also physically plausible. 

Maybe you want to remove the keywords `PLYSHLOG` in `POLY.inc`, because shear calculation really depends on the field size. When you change the field a lot, it will not be physically plausible without updating `PLYSHLOG` setup. 

I have not tested a lot for other cases. My suggestion is to try to make the case physically viable, which is the safe way to do.

**Comment by Wenyue-Sun** [comment]
> Hi, Wenyue.
> 
> Sorry for not getting back to you sooner. I took a brief look of your example 2D_OILWATER_POLYMER_USE_PORV.DATA .
> 
> In this example, the cell volume is 1, and you assigned pore volume 50000, which will results in porosity of 50000. It depends how the simulator is formulated, it might or might not cause big problem. It depends in details how your equation is formulated and how the geometry information is used. It causes problem for OPM-flow, but I could not find enough time to dig in for this.
> 
> For this example, I suggest you increase the grid cell volume, then make sure the pore volume is smaller than the grid cell volume, so you have a porosity smaller than 1, which is also physically plausible.
> 
> Maybe you want to remove the keywords `PLYSHLOG` in `POLY.inc`, because shear calculation really depends on the field size. When you change the field a lot, it will not be physically plausible without updating `PLYSHLOG` setup.
> 
> I have not tested a lot for other cases. My suggestion is to try to make the case physically viable, which is the safe way to do.

Got it. Thanks, did notice that the convergence issue go away with DX/DY/DZ being physical numbers that lead to porosity < 1. Good to know the reason. However, the issue that polymer does not go with NNC still exists, not sure what is the root cause here. 

Appreciate your prompt response!

**Comment by GitPaean** [comment]
> Got it. Thanks, did notice that the convergence issue go away with DX/DY/DZ being physical numbers that lead to porosity < 1. Good to know the reason. However, the issue that polymer does not go with NNC still exists, not sure what is the root cause here.

> Appreciate your prompt response!

Hi, I did not have any problem in running the example 2D_OILWATER_POLYMER_USE_NNC.DATA. 

```
Number of MPI processes:         1
Threads per MPI process:         2
Total time (seconds):            0.72 
Solver time (seconds):           0.71 
 Assembly time (seconds):        0.27 (Failed: 0.0; 0.0%)
   Well assembly (seconds):      0.13 (Failed: 0.0; 0.0%)
 Linear solve time (seconds):    0.06 (Failed: 0.0; 0.0%)
   Linear setup (seconds):       0.02 (Failed: 0.0; 0.0%)
 Update time (seconds):          0.08 (Failed: 0.0; 0.0%)
 Pre/post step (seconds):        0.12 (Failed: 0.0; 0.0%)
 Output write time (seconds):    0.17
Overall Linearizations:        968    (Failed:   0; 0.0%)
Overall Newton Iterations:     832    (Failed:   0; 0.0%)
Overall Linear Iterations:    2367    (Failed:   0; 0.0%)
```

I am using the master branch. 

And also, if you use NNC, I am not sure how the NNC interact with the shear part. I encourage you to remove the keyword PLYSHLOG from the POLY.inc, which is one of the keywords use the geometry information the most and I am not sure we handle all the situations well, depending what kind of cases we handled before.

**Comment by Wenyue-Sun** [comment]
> > Got it. Thanks, did notice that the convergence issue go away with DX/DY/DZ being physical numbers that lead to porosity < 1. Good to know the reason. However, the issue that polymer does not go with NNC still exists, not sure what is the root cause here.
> 
> > Appreciate your prompt response!
> 
> Hi, I did not have any problem in running the example 2D_OILWATER_POLYMER_USE_NNC.DATA.
> 
> ```
> Number of MPI processes:         1
> Threads per MPI process:         2
> Total time (seconds):            0.72 
> Solver time (seconds):           0.71 
>  Assembly time (seconds):        0.27 (Failed: 0.0; 0.0%)
>    Well assembly (seconds):      0.13 (Failed: 0.0; 0.0%)
>  Linear solve time (seconds):    0.06 (Failed: 0.0; 0.0%)
>    Linear setup (seconds):       0.02 (Failed: 0.0; 0.0%)
>  Update time (seconds):          0.08 (Failed: 0.0; 0.0%)
>  Pre/post step (seconds):        0.12 (Failed: 0.0; 0.0%)
>  Output write time (seconds):    0.17
> Overall Linearizations:        968    (Failed:   0; 0.0%)
> Overall Newton Iterations:     832    (Failed:   0; 0.0%)
> Overall Linear Iterations:    2367    (Failed:   0; 0.0%)
> ```
> 
> I am using the master branch.
> 
> And also, if you use NNC, I am not sure how the NNC interact with the shear part. I encourage you to remove the keyword PLYSHLOG from the POLY.inc, which is one of the keywords use the geometry information the most and I am not sure we handle all the situations well, depending what kind of cases we handled before.

Hi, Kai, 

Thanks for the reply. For the "2D_OILWATER_POLYMER_USE_NNC.DATA" case, it actually converged. The issue I meant was that the final polymer concentration map is not correct (being zero). (I turned off the PLYSHLOG keyword). The results figure I got from OPM is the following: 
![image](https://user-images.githubusercontent.com/17682458/123459059-670d9780-d5ab-11eb-80ef-024edc440fc6.png)
, though the pressure map looks 'right': 
![image](https://user-images.githubusercontent.com/17682458/123459264-a9cf6f80-d5ab-11eb-82d2-b4ba027b0cbe.png)


With the exact simulation deck (nothing changed), I tried a commercial simulator, the returned result for polymer concentration is like:
![2D_OILWATER_POLYMER_USE_NNC](https://user-images.githubusercontent.com/17682458/123459168-8c9aa100-d5ab-11eb-90d5-7ee4a2bb70a9.JPG)

So basically the polymer should be moving based on the Non-neighbor connection (NNC) that is specified in the simulation deck (that layer 5 only connects to layer 3 and layer 1), which I did not observe in the OPM simulation result currently. 

Not sure whether this clarifies the issue. Thanks for your attention.
